server/src/pull/next-page.ts

```diff
@@ -0,0 +1,189 @@
+    LIMIT
+  );
+}
+
+export async function readNextPage(
+  executor: Executor,
+  clientGroupID: string,
+  order: number,
+) {
+  let remaining = LIMIT;
+  const issues: Issue[] = [];
+  const descriptions: Description[] = [];
+  const comments: Comment[] = [];
+
+  const issueDeletes: string[] = [];
+  const descriptionDeletes: string[] = [];
+  const commentDeletes: string[] = [];
+
+  const issueMeta: SearchResult[] = [];
+  const descriptionMeta: SearchResult[] = [];
+  const commentMeta: SearchResult[] = [];
+
+  function buildReturn() {
+    return {
+      issues,
+      descriptions,
+      comments,
+      issueDeletes,
+      descriptionDeletes,
+      commentDeletes,
+      issueMeta,
+      descriptionMeta,
+      commentMeta,
+    };
+  }
+
+  // TODO: optimize to not require 3 queries in turn.
+  // Issue all of them at once and throw away results over limit?
+  // Issue a query just for `ids` and `versions` via a union then fulfill a page of data?
+  const {rows: issueRows} = await findUnsentItems(
+    executor,
+    'issue',
+    clientGroupID,
+    order,
+    remaining,
+  );
+  for (const r of issueRows) {
+    issueMeta.push({
+      id: r.id,
+      rowversion: r.rowversion,
+    });
+    issues.push({
+      id: r.id,
+      title: r.title,
+      priority: r.priority,
+      status: r.status,
+      modified: parseInt(r.modified),
+      created: parseInt(r.created),
+      creator: r.creator,
+      kanbanOrder: r.kanbanorder,
+    });
+  }
+
+  remaining -= issueRows.length;
+  if (remaining <= 0) {
+    return buildReturn();
+  }
+
+  const {rows: descriptionRows} = await findUnsentItems(
+    executor,
+    'description',
+    clientGroupID,
+    order,
+    remaining,
+  );
+  for (const r of descriptionRows) {
+    descriptionMeta.push({
+      id: r.id,
+      rowversion: r.rowversion,
+    });
+    descriptions.push({
+      id: r.id,
+      body: r.body,
+    });
+  }
+
+  remaining -= descriptionRows.length;
+  if (remaining <= 0) {
+    return buildReturn();
+  }
+
+  const {rows: commentRows} = await findUnsentItems(
+    executor,
+    'comment',
+    clientGroupID,
+    order,
+    remaining,
+  );
+  for (const r of commentRows) {
+    commentMeta.push({
+      id: r.id,
+      rowversion: r.rowversion,
+    });
+    comments.push({
+      id: r.id,
+      issueID: r.issueid,
+      created: parseInt(r.created),
+      body: r.body,
+      creator: r.creator,
+    });
+  }
+
+  remaining -= commentRows.length;
+  if (remaining <= 0) {
+    return buildReturn();
+  }
+
+  const [
+    {rows: issueDeleteRows},
+    {rows: descriptionDeleteRows},
+    {rows: commentDeleteRows},
+  ] = await Promise.all([
+    findDeletions(executor, 'issue', clientGroupID, order, remaining),
+    findDeletions(executor, 'description', clientGroupID, order, remaining),
+    findDeletions(executor, 'comment', clientGroupID, order, remaining),
+  ]);
+
+  // Record deletes against the _current_ CVR
```

Comment:
This sounds important but I'm not getting the distinction. The `findUnsentItems` function finds rows that are new. Doesn't this do the same for deletions that are new?

Reply:
This was a comment from a prior iteration where I was thinking through ideas to be able to remove tombstones.

Will remove the comment.

Reply:
resolved ddee4f1ad4ef54fbbf868b2940a397b4069b642e